tests/common/devices/multi_asic.py

```diff
@@ -614,3 +620,20 @@ def docker_copy_from_asic(self, container_name, src, dst, asic_id = 0):
         if duthost.is_multi_asic:
             container_name += str(asic_id)
         self.shell("sudo docker cp {}:{} {}".format(container_name, src, dst))
+        
+    def is_service_fully_started_per_asic_or_host(self, service):
+        """This function tell if service is fully started base on multi-asic/single-asic"""
+        duthost = self.sonichost
+        if duthost.is_multi_asic:
+            for asic_index in range(duthost.facts["num_asic"]):
+                docker_name = self.asic_instance(asic_index).get_docker_name(service)
```

Comment:
There could be a situation when asic_index is not present in ASICS_PRESENT, then self.asic_instance(asic_index) will not be present.
for loop can be changes to look at asics in self.asics ?
Do we want to check the status of a service for asic that is not present also?
@arlakshm - can you suggest?

Reply:
Looks like we should not use 'range(num_asics)', but just iterate through the asics list itself.

Reply:
can you create seprate PR for this change ? let's keep this pr to handle ASIC_PRESENT only

Reply:
I think they need to go together in the same PR - as it breaks tests/functionality.

Here is where 'num_asic' is used in tests directory.

```
(~) > grep -r "[\"']num_asic[\"']"
common/devices/multi_asic.py:                if self.sonichost.facts['num_asic'] == 1:
common/devices/multi_asic.py:            if self.sonichost.facts['num_asic'] == 1:
common/devices/multi_asic.py:        if self.sonichost.facts['num_asic'] == 1:
common/devices/multi_asic.py:        if self.sonichost.facts['num_asic'] == 1:
common/devices/multi_asic.py:        if self.sonichost.facts['num_asic'] == 1:
common/devices/multi_asic.py:        if self.sonichost.facts['num_asic'] == 1:
common/devices/multi_asic.py:        if self.sonichost.facts['num_asic'] == 1:
common/devices/multi_asic.py:        if self.sonichost.facts['num_asic'] == 1:
common/devices/multi_asic.py:        if self.sonichost.facts['num_asic'] == 1:
common/devices/multi_asic.py:        if self.sonichost.facts['num_asic'] == 1 or namespace == DEFAULT_NAMESPACE:
common/devices/multi_asic.py:        if self.sonichost.facts['num_asic'] == 1 or asic_id == DEFAULT_ASIC_ID:
common/devices/multi_asic.py:            for n in range(duthost.facts['num_asic']):
common/devices/multi_asic.py:            for n in range(duthost.facts['num_asic']):
common/devices/multi_asic.py:            for asic_index in range(duthost.facts["num_asic"]):
common/devices/sonic.py:        self.is_multi_asic = True if self.facts["num_asic"] > 1 else False
common/devices/sonic.py:                "num_asic": 1,
common/devices/sonic.py:        facts["num_asic"] = self._get_asic_count(facts["platform"])
common/devices/sonic.py:        facts["asics_present"] = asics_present if len(asics_present) != 0 else range(facts["num_asic"])
common/devices/sonic.py:        return self.facts["num_asic"]
common/devices/sonic.py:        if self.facts["num_asic"] == 1:
common/devices/sonic.py:            num_asics = int(self.facts["num_asic"])
common/devices/sonic.py:        if self.facts["num_asic"] == 1:
common/devices/sonic.py:            num_asics = self.facts["num_asic"]
common/devices/sonic.py:        num_asics = int(self.facts["num_asic"])
common/devices/sonic_asic.py:        if self.sonichost.facts['num_asic'] != 1:
common/plugins/sanity_check/checks.py:                        if dut.facts['num_asic'] == 1:
common/plugins/sanity_check/checks.py:                        if dut.facts['num_asic'] == 1:
common/plugins/conditional_mark/README.md:      "num_asic": 1,
voq/voq_helpers.py:    if host.facts["num_asic"] > 1:
voq/voq_helpers.py:    if host.facts["num_asic"] == 1:
voq/voq_helpers.py:    if host.facts["num_asic"] == 1:
voq/voq_helpers.py:    if rem_host.facts["num_asic"] > 1:
voq/test_voq_ipfwd.py:        if ports['portA']['dut'].get_facts()['num_asic'] > 1:
voq/test_voq_ipfwd.py:            if ports['portC']['dut'].get_facts()['num_asic'] > 1:
voq/test_voq_ipfwd.py:        if ports['portA']['dut'].get_facts()['num_asic'] == 1:
voq/test_voq_ipfwd.py:            if ports['portD']['dut'].get_facts()['num_asic'] > 1:
voq/test_voq_ipfwd.py:        if ports['portA']['dut'].get_facts()['num_asic'] == 1:
voq/test_voq_ipfwd.py:        if ports['portA']['dut'].get_facts()['num_asic'] == 1:
voq/test_voq_ipfwd.py:    if host.facts['num_asic'] > 1:
voq/test_voq_ipfwd.py:            if ports['portA']['dut'].get_facts()['num_asic'] > 1:

```

The changes are limited to multi_asic.py only.   Other use of 'num_asic' is primarily to check if the DUT has more than 1 asic. So, it is not a big change.

Reply:
@abdosi  - are we ok to make the change to replace range with iterating over the self.asics in the 3 places in multi_asic.py - if so, we can push a commit for the changes.

Reply:
similar change to use list of selected asics should be done in docker_cmds_on_all_asics() and docker_copy_to_all_asics() also.